### ISSUE TITLE
Update QPH storage mechanism for step-18 and step-44

### ISSUE THREAD

**Issue by jppelteret** [dealii/dealii#1438]
As per my discussion with @tamiko , we should update step-18 and step-44 to store the QPH in a map as opposed to a global vector. I'll undertake to do this and open a PR in the coming weeks.

**Comment by bangerth** [comment]
What was the rationale for doing so?

**Comment by bangerth** [comment]
(Not opposed on principle, but curious?)

**Comment by jppelteret** [comment]
It just seemed like a more elegant approach (and perhaps safer?) as it doesn't require that one store data using the `cell->user_pointer()`, but rather with the cell iterator as the key in the map. For one, we can dynamically change the map's size as we create or remove QPH objects  (well, its easier to do than if we use this global vector) and we can remove all instances of `reinterpret_cast` as well. 

So we could have, for example, an interface such as this (shameless copy-paste from another mini-project, so the underlying mechanics are of course up for discussion):
```c++
    void
    add_cell_lqph(const QPH_DOF_CELL_TYPE & cell)
    {
      quadrature_point_history.insert(std::make_pair(static_cast<QPH_TRIA_CELL_TYPE>(cell),
                                                     std::vector<PointHistory<dim> > (n_q_points)) );
    }

    QPH_STORAGE &
    get_cell_lqph (const QPH_TRIA_CELL_TYPE & cell)
    {
        const typename MAP_QPH_STORAGE::iterator it = quadrature_point_history.find(cell);
        AssertThrow(it != quadrature_point_history.end(),
                    ExcMessage("QPH not found for this cell."));
        return it->second;
    }

    PointHistory<dim> &
    get_lqph (const QPH_DOF_CELL_TYPE & cell,
              const unsigned int q_point)
    {
        return get_cell_lqph(cell)[q_point];
    }
```

And the initial set-up call could be simplified to the following:
```c++
template <int dim>
  void Solid<dim>::setup_qph(const bool clear_data)
  {
    std::cout << "    Setting up quadrature point data..." << std::endl;

    if (clear_data == true)
      quadrature_point_history.clear();

    // Next we setup the initial quadrature
    // point data:
    for (typename DoFHandler<dim>::active_cell_iterator cell =
           dof_handler_ref.begin_active(); cell != dof_handler_ref.end();
         ++cell)
      {
        add_cell_lqph(cell);

        QPH_STORAGE &lqph = get_cell_lqph(cell);

        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
            lqph[q_point].setup_lqp(parameters);
      }
  }
```
Assuming that one day we wrote a QPH data manager class, this sort of functionality would be incorporated into it.

**Comment by bangerth** [comment]
> So we could have, for example, an interface such as this (shameless
> copy-paste from another mini-project, so the underlying mechanics are of
> course up for discussion):

Yes, this is actually quite nice -- I like it :-)


>      void
>      add_cell_lqph(const  QPH_DOF_CELL_TYPE & cell)
>      {
>        quadrature_point_history.insert(std::make_pair(static_cast<QPH_TRIA_CELL_TYPE>(cell),
>                                                       std::vector<PointHistory<dim> > (n_q_points)) );

You can also save the `static_cast` here if you simply do
```
   quadrature_point_history[cell] = std::vector<PointHistory<dim> > 
(n_q_points);
```

**Comment by jppelteret** [comment]
Great! I'll put together the concept in a couple of weeks (I'm away travelling for a short while). I'll let the discussion brew until then and set up a PR when I have something sensible to show that takes everyone's points into consideration :-)

**Comment by jppelteret** [comment]
Fixed by #2620
